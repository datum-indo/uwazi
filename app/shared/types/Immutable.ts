/** @format */

export type IImmutable<T> = T extends string
  ? string
  : T extends number
  ? number
  : T extends boolean
  ? boolean
  : T extends null
  ? null
  : T extends undefined
  ? undefined
  : T extends Array<infer Elem>
  ? {
      toJS(): Elem[];
      get(i: number): IImmutable<Elem>;
      findIndex(fn: (e: IImmutable<Elem>) => boolean): number;
      filter(fn: (e: IImmutable<Elem>) => boolean): IImmutable<T>;
      find(fn: (e: IImmutable<Elem>) => boolean): IImmutable<Elem> | undefined;
      map<T2>(fn: (e: IImmutable<Elem>) => T2): IImmutable<T2[]>;
      reduce<R>(fn: (r: R, e: IImmutable<Elem>) => R, s: R): R;
    }
  : {
      toJS(): T;
      get<Field extends keyof T>(_field: Field): IImmutable<T[Field]>;
    };
