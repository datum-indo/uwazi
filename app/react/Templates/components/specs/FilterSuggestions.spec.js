import React from 'react';
import { shallow } from 'enzyme';
import Immutable from 'immutable';

import { FilterSuggestions } from '../FilterSuggestions';

describe('FilterSuggestions', () => {
  let component;
  let props;
  let templates;
  let thesauris;

  function renderComponent(label = 'test', type = 'text', content) {
    templates = [
      {
        _id: 'template1',
        properties: [
          { localID: 1, label, filter: true, type },
          { localID: 2, label: 'something else' },
        ],
      },
      {
        _id: 'template2',
        name: 'Template 2',
        properties: [
          { label: 'Date', type: 'date', filter: true },
          { label: 'Author', type: 'text', filter: true },
          { label: 'filterFalse', type: 'text', filter: false },
          { label: 'Authors', type: 'select', filter: true, content: 'abc1' },
        ],
      },
      {
        _id: 'template3',
        name: 'Template 3',
        properties: [
          { label: 'date ', type: 'date', filter: true },
          { label: 'filterFalse', type: 'text', filter: true },
          { label: 'Keywords', type: 'text', filter: true },
        ],
      },
    ];

    thesauris = [
      { _id: 'abc1', name: 'Best SCI FI Authors' },
      { _id: 'abc2', name: 'Favourite dessert recipes' },
    ];

    props = {
      label,
      type,
      filter: true,
      content,
      data: { name: 'Current template', _id: 'template1' },
      templates: Immutable.fromJS(templates),
      thesauris: Immutable.fromJS(thesauris),
    };

    component = shallow(<FilterSuggestions {...props} />);
  }

  it('should always render the current property as a guide', () => {
    renderComponent('Year', 'date');
    const suggestion = component.find('tbody > tr').at(0);
    expect(suggestion).toMatchSnapshot();
  });

  describe('when matches type and label as other template property', () => {
    it('should show a message', () => {
      renderComponent('author', 'text');
      const suggestion = component.find('tbody > tr').at(1);
      expect(suggestion).toMatchSnapshot();
    });
  });

  describe('when label is the same but different type', () => {
    it('should mark it as conflict', () => {
      renderComponent('author', 'date');
      const suggestion = component.find('.conflict');
      expect(suggestion).toMatchSnapshot();
    });
  });

  describe('when label is the same but different content', () => {
    it('should mark it as conflict', () => {
      renderComponent('authors', 'select', 'abc2');
      const suggestion = component.find('.conflict');
      expect(suggestion).toMatchSnapshot();
    });
  });

  describe('when content does not match with a thesauri (happening when opening property and the dictionary has been deleted)', () => {
    it('should not show anything', () => {
      renderComponent('authors', 'select', 'non existent thesauri');
      const suggestion = component.find('tbody > tr').at(1);
      expect(suggestion).toMatchSnapshot();
    });
  });

  describe('when props.filter = flase', () => {
    it('should now show the preoperty with same name', () => {
      renderComponent('filterFalse', 'text');
      const suggestion = component.find('tbody > tr').at(1);
      expect(suggestion).toMatchSnapshot();
    });
  });
});
