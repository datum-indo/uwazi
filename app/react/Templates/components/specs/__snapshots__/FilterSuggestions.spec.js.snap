// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSuggestions should always render the current property as a guide 1`] = `
<tr
  className="property-atributes is-active"
  title="This is the current property and will be used together with equal properties."
>
  <td>
    <Connect(Icon)
      icon="file"
    />
     
    Current template
  </td>
  <td>
    <Connect(Icon)
      icon="calendar"
    />
     
    Date
  </td>
</tr>
`;

exports[`FilterSuggestions when content does not match with a thesauri (happening when opening property and the dictionary has been deleted) should not show anything 1`] = `
<tr
  className="property-atributes is-active"
  key="0"
  title="This property has different Thesauri and wont't be used together."
>
  <td>
    <Connect(Icon)
      icon="file"
    />
     
    Template 2
  </td>
  <td
    className=""
  >
    <Connect(Icon)
      icon="warning"
    />
    <Connect(Icon)
      icon="sort"
    />
     
    Select
  </td>
  <td
    className="conflict"
  >
    <Connect(Icon)
      icon="warning"
    />
    <Connect(Icon)
      icon="book"
    />
     
    Best SCI FI Authors
  </td>
</tr>
`;

exports[`FilterSuggestions when label is the same but different content should mark it as conflict 1`] = `
<td
  className="conflict"
>
  <Connect(Icon)
    icon="warning"
  />
  <Connect(Icon)
    icon="book"
  />
   
  Best SCI FI Authors
</td>
`;

exports[`FilterSuggestions when label is the same but different type should mark it as conflict 1`] = `
<td
  className="conflict"
>
  <Connect(Icon)
    icon="warning"
  />
  <Connect(Icon)
    icon="font"
  />
   
  Text
</td>
`;

exports[`FilterSuggestions when matches type and label as other template property should show a message 1`] = `
<tr
  className="property-atributes is-active"
  key="0"
  title="This property has the same configuration and will be used together."
>
  <td>
    <Connect(Icon)
      icon="file"
    />
     
    Template 2
  </td>
  <td
    className=""
  >
    <Connect(Icon)
      icon="warning"
    />
    <Connect(Icon)
      icon="font"
    />
     
    Text
  </td>
</tr>
`;

exports[`FilterSuggestions when props.filter = flase should now show the preoperty with same name 1`] = `
<tr
  className="property-atributes is-active"
  key="0"
  title="This property has the same configuration and will be used together."
>
  <td>
    <Connect(Icon)
      icon="file"
    />
     
    Template 3
  </td>
  <td
    className=""
  >
    <Connect(Icon)
      icon="warning"
    />
    <Connect(Icon)
      icon="font"
    />
     
    Text
  </td>
</tr>
`;
